modules/RTC/TraceablePeerConnection.js

```diff
@@ -1481,21 +1481,23 @@ TraceablePeerConnection.prototype.containsTrack = function(track) {
 /**
  * Add {@link JitsiLocalTrack} to this TPC.
  * @param {JitsiLocalTrack} track
+ * @returns {Promise<void>} - resolved when done.
```

Comment:
We have made `addTrack` return a Promise now but in JingleSessionPC.js where the local tracks are added before offer/answer, we are not waiting for the promise to get resolved.

Reply:
I'll take a look - must have missed it, thanks!

Reply:
It seems that the encoding could be passed as `transceiverInit`[1] and `addTrack` won't need it do be async? I'm wondering why is it currently done only for the "isInitiator and not Firefox" case? Is it possible to get rid of [2] and pass the encoding to the constructor?

Also I don't understand why TCPUtils makes assumptions about whether or not tracks are added before 'session-initiate' is sent to another peer[3]. Dealing with problems like this should belong to the jingle session layer and should be of no concern to TCPUtils which belong the RTC layer.

[1]: https://github.com/jitsi/lib-jitsi-meet/blob/d219b6b91b60c6f726b66576f34cb67adab23abd/modules/RTC/TPCUtils.js#L225
[2]: https://github.com/jitsi/lib-jitsi-meet/blob/d219b6b91b60c6f726b66576f34cb67adab23abd/modules/RTC/TraceablePeerConnection.js#L1547
[3]: https://github.com/jitsi/lib-jitsi-meet/blob/d219b6b91b60c6f726b66576f34cb67adab23abd/modules/RTC/TPCUtils.js#L216

Reply:
FF doesn't take the encodings in the `transceiverInit`, they have to be configured after the transceiver is created by calling `setParameters` on the sender. That is the major difference between Chrome/Safari unified vs Firefox unified plan impl. 

In the transceiver model, we do have to determine how to add the tracks based on whether the client is an initiator or a responder. In the responder case, a transceiver is already created when we call sRD. Calling addTransceiver() again would create a new transceiver, so pc.addTrack() needs to be called in this case to make sure the track gets attached to the existing transceiver.
The RTC layer is what add the tracks to the pc, so these decisions have to be made here.

Reply:
When I'm looking at the code in `JingleSessionPC.setOfferAnswerCycle` the tracks are added, before the remote description is set.

If RTC module relies on whether or not a transceiver exists it should do a check for it by scanning the list on the peerconnection. Same applies to relying on whether or not a remote description has been set - the code should check the `remoteDescription` field.

Reply:
Makes sense, as discussed, we can clean this up when working on p2p case in unified since this code path is not being used in the jvb case which is the only active mode for unified plan now.